machine/xen/v1alpha1.go

```diff
@@ -0,0 +1,536 @@
+		machine, err := service.Get(ctx, &machine)
+		if err != nil {
+			machines.Items = cached
+			return machines, err
+		}
+
+		machines.Items[i] = *machine
+	}
+
+	return machines, nil
+}
+
+func (service *machineV1alpha1Service) Watch(ctx context.Context, machine *machinev1alpha1.Machine) (chan *machinev1alpha1.Machine, chan error, error) {
+	domId, ok := machine.Status.PlatformConfig.(xenlight.Domid)
+	if !ok {
+		return nil, nil, fmt.Errorf("machine has no platform config")
+	}
+
+	w, err := NewWatcher(domId)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// signals when xenstore tree was updated for domain
+	watch, err := w.Watch(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	events := make(chan *machinev1alpha1.Machine)
+	errs := make(chan error)
+
+	go func() {
+		intialMachine, err := service.Get(ctx, machine)
+		if err != nil {
+			errs <- err
+		}
+		events <- intialMachine
+
+		for {
+			select {
+			case <-ctx.Done():
+				w.Close()
+				return
+			case <-watch:
+				machine, err := service.Get(ctx, machine)
+				if err != nil {
+					errs <- err
+					continue
+				}
+
+				events <- machine
+			}
+		}
+	}()
+
+	return events, errs, nil
+}
+func (service *machineV1alpha1Service) Logs(ctx context.Context, machine *machinev1alpha1.Machine) (chan string, chan error, error) {
+	domId, ok := machine.Status.PlatformConfig.(xenlight.Domid)
+	if !ok {
+		return nil, nil, fmt.Errorf("machine has no platform config")
+	}
+
+	xenCtx, err := xenlight.NewContext()
+	if err != nil {
+		return nil, nil, fmt.Errorf("could not create xen context: %v", err)
+	}
+	defer xenCtx.Close()
+
+	pts, err := xenCtx.PrimaryConsoleGetTty(uint32(domId))
+	if err != nil {
+		return nil, nil, fmt.Errorf("could not get xen domain pts: %v", err)
+	}
+
+	// Start appending pts output to logfile: pts -> chan -> log file
+	go func() {
+		ptsChan := make(chan []byte)
+		errChan := make(chan error)
+
+		ptsFD, err := os.OpenFile(pts, os.O_RDONLY, 0o644)
+		if err != nil {
+			log.G(ctx).Errorf("could not open xen domain pts: %v", err)
+			return
+		}
+
+		go func() {
+			buf := make([]byte, 1024)
+			for {
+				n, err := ptsFD.Read(buf)
+				if err != nil {
+					if err != os.ErrClosed {
+						errChan <- err
+					}
+					return
+				}
+				ptsChan <- buf[:n]
+			}
+		}()
+
+		logFD, err := os.OpenFile(machine.Status.LogFile, os.O_APPEND|os.O_WRONLY, 0o644)
+		if err != nil {
+			log.G(ctx).Errorf("log file not found after create: %v", err)
+			return
+		}
+
+		for {
+			select {
+			case err := <-errChan:
+				log.G(ctx).Errorf("could not read from pts: %v", err)
+			case line := <-ptsChan:
+				_, err := logFD.Write(line)
+				if err != nil {
+					log.G(ctx).Errorf("could not write to log file: %v", err)
+				}
+			case <-ctx.Done():
+				logFD.Close()
+				ptsFD.Close()
+				return
+			}
+		}
+	}()
+
+	return logtail.NewLogTail(ctx, machine.Status.LogFile)
+}
+
+func getXenState(domInfo *xenlight.Dominfo) (int, machinev1alpha1.MachineState) {
```

Comment:
Add a comment explaining `-1/0/1`, or even better turn the `int` into an enum `exitStatus` or something like that

`type exitStatus int`

Reply:
To be honest i am not sure about the spec for exit codes. These are the same as for `qemu`.

Reply:
sounds good then